### PR TITLE
Oracles: Move oracle Flag to an attribute

### PR DESCRIPTION
**Pull request by shargon**
Wait for https://github.com/neo-project/neo/pull/1620
Related to the conversation https://github.com/neo-project/neo/pull/1609

### CODE REVIEW

File: src/neo/Network/P2P/Payloads/Transaction.cs [hunk 1]
```diff
@@ -39,11 +39,11 @@ public class Transaction : IEquatable<Transaction>, IInventory, IInteroperable
         private TransactionAttribute[] attributes;
```
Comment (Tommo-L) [inline src/neo/Network/P2P/Payloads/Transaction.cs:39]: Is one attribute enough? @erikzhang what do you think?

Comment (shargon) [reply src/neo/Network/P2P/Payloads/Transaction.cs:39]: I think that if we have attributes, it should be an array.

File: src/neo/Network/P2P/Payloads/Transaction.cs [hunk 2]
```diff
@@ -346,6 +318,45 @@ public virtual VerifyResult Verify(StoreView snapshot, BigInteger totalSenderFee
             return VerifyResult.Succeed;
         }
 
+        #region Oracles
+
+        private void CheckOracleAttribute()
```
Comment (Tommo-L) [inline src/neo/Network/P2P/Payloads/Transaction.cs:323]: Change void to bool? Or we should rename it `LoadOracleAttribute`?

File: src/neo/Network/P2P/Payloads/OracleAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
+using Neo.IO;
+using System;
+using System.IO;
+
+namespace Neo.Network.P2P.Payloads
+{
+    public class OracleAttribute : ISerializable
+    {
+        public enum OracleAttributeType : byte
```
Comment (erikzhang) [inline src/neo/Network/P2P/Payloads/OracleAttribute.cs:9]: Move to `TransactionAttributeUsage`?

**Comment by erikzhang** [comment]
Should we move cosigners to attributes too?

**Comment by shargon** [comment]
> Should we move cosigners to attributes too?

It's possible, but it's slower, you need to cache it, renew the cache ...

File: src/neo/Network/P2P/Payloads/Transaction.cs [hunk 4]
```diff
@@ -346,6 +334,57 @@ public virtual VerifyResult Verify(StoreView snapshot, BigInteger totalSenderFee
...
+            }
+            else
+            {
+                oracleResponse = null;
+            }
+        }
+
+        public bool IsOracleResponse(out UInt256 requestTx)
+        {
+            LoadOracleAttribute();
```
Comment (Tommo-L) [inline src/neo/Network/P2P/Payloads/Transaction.cs:368]: Move `LoadOracleAttribute` into `Deserialize`?

Comment (shargon) [reply src/neo/Network/P2P/Payloads/Transaction.cs:368]: We can call it in Deserialize, but it could be changed outside

**Comment by erikzhang** [comment]
> It's possible, but it's slower, you need to cache it, renew the cache ...

What do you mean?

File: src/neo/Network/P2P/Payloads/Transaction.cs [hunk 5]
```diff
@@ -346,6 +334,57 @@ public virtual VerifyResult Verify(StoreView snapshot, BigInteger totalSenderFee
...
+
+            if (attr != null)
+            {
+                oracleCached |= OracleCache.Response;
+                oracleResponse = attr.Data.AsSerializable<OracleResponseAttribute>();
+            }
+            else
+            {
+                oracleResponse = null;
+            }
```
Comment (shargon) [inline src/neo/Network/P2P/Payloads/Transaction.cs:363]: If you need to check the cosigners, you need to fetch all the attributes and find/parse the expected type. Is the same logic that I need for check if it's a oracle request, or oracle response or both.

Comment (shargon) [reply src/neo/Network/P2P/Payloads/Transaction.cs:363]: @erikzhang please take a look to https://github.com/neo-project/neo/pull/1620

**Comment by shargon** [comment]
Refactorized and ready to review again

**Pull request merged**